File: OSVParser/parser.js
var pg = require('pg');
//or native libpq bindings
//var pg = require('pg').native

var conString = "postgres://postgres:Asdfjkl1@localhost/gis";

var client = new pg.Client(conString);
client.connect(function(err) {
  if(err) {
    return console.error('could not connect to postgres', err);
  }
  var sql = "declare geo no scroll cursor with hold for "
  sql += "SELECT row_to_json(fc) from (select ST_AsGeoJSON(lg.way)::json As geometry "
  sql += ", row_to_json(lp) As properties "
  sql += "FROM planet_osm_line As lg "
  sql += "INNER JOIN (SELECT osm_id, name,highway FROM planet_osm_line where highway is not null) As lp "
  sql += "ON lg.osm_id = lp.osm_id) as fc;"
//sql = "select * from planet_osm_line limit 1"
  //console.log(sql)
  var num = 0;
  client.query(sql, function(err, result) {
    if(err) {
      return console.error('error running query', err);
    }
  })
  sql = "fetch forward 50 from geo "
  var work = function(){
  client.query(sql, function(err, result) {
  	var count = result.rowCount
  	num++
 	console.log(num)
  	if(!count || count == 0)
  	{
  		clearInterval(timer)
  		return
  	}
  	console.log(result.rowCount)
    for(i=0;i<count;i++)
    {
    	//console.log(result.rows[i].row_to_json.geometry.coordinates);
    	//console.log(result.rows[i].row_to_json.properties)
    		var tempchunks = coordArrayToChunks2(parseLinestringRoadtoLineCoords(result.rows[i].row_to_json.geometry.coordinates),result.rows[i].row_to_json.properties.name)
	//todo: return one chunk and pass it to mongo
			for (var c in tempchunks)//check each chunk that was just spit out from processing functions....
		{
			writeMongo(c)
		}
	}
	//parseOSMtoVoxeljs(result).length
	console.log(count)
	timer = setTimeout(work,100);
    //output: Tue Jan 15 2013 19:12:47 GMT-600 (CST)
    //client.end();
  });
  }
  timer = setTimeout(work,2000);
});

//^^^^^^^^^^^^^^^^^^^^OSM QUERIES^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
//vvvvvvvvvvvvvvvvvvOSM to Chunk calculations =vvvvvvvvvvvvvvvvv


//Convert coords the use the equator to xy of an onscreen map for continous coords.
//hard values are max 3857 values.  IF is not needed, the math works that both 
//+ and - coords convert correctly with the same formula.
//Math.floor because I don't care about percents of a meter.
function convertEPSG3857toScreenXY(lat,lon)
{
	var xlat,ylon
	xlat = 20026376 + Math.floor(lat);
	ylon = 20048966 - Math.floor(lon);
	//console.log(lat,lon);
	return {x:xlat,y:ylon};
	//return [lat,lon];
}

//Code from 
//http://stackoverflow.com/questions/4672279/bresenham-algorithm-in-javascript
function calcStraightLine (startCoordinates, endCoordinates) 
{
    var coordinatesArray = new Array();
    // Translate coordinates
    var x1 = startCoordinates.x;
    var y1 = startCoordinates.y;
    var x2 = endCoordinates.x
    var y2 = endCoordinates.y;
    // Define differences and error check
    var dx = Math.abs(x2 - x1);
    var dy = Math.abs(y2 - y1);
    var sx = (x1 < x2) ? 1 : -1;
    var sy = (y1 < y2) ? 1 : -1;
    var err = dx - dy;
    // Set first coordinates
    coordinatesArray.push({y:y1, x:x1});
    //return coordinatesArray;
    // Main loop
    while (!((x1 == x2) && (y1 == y2))) {
      var e2 = err << 1;
      if (e2 > -dy) {
        err -= dy;
        x1 += sx;
      }
      if (e2 < dx) {
        err += dx;
        y1 += sy;
      }
      // Set coordinates
      coordinatesArray.push({y:y1, x:x1});
      //console.log(x1,'+',y1);
    }
    // Return the result
    return coordinatesArray;
}


//Loop through linestring with points that should form a line.
//convert each point from gps to xy
//form line of xy's between each xy.
//combine all those lines together.
function parseLinestringRoadtoLineCoords(LineStringArray)
{
	var lineCoordsArray = new Array();
	for (i = 0;i < LineStringArray.length - 1;i++)
	{
		lineCoordsArray = lineCoordsArray.concat(
							calcStraightLine(
								convertEPSG3857toScreenXY(LineStringArray[i][0],LineStringArray[i][1]),
								convertEPSG3857toScreenXY(LineStringArray[i+1][0],LineStringArray[i+1][1])
								)
							)
	}
	//console.log(lineCoordsArray);
	return lineCoordsArray;
}


function calcChunkNameFromCoords(xlat,ylon,dims)
{
	return Math.floor(xlat / (dims+1)) + '|0|' + Math.floor(ylon / (dims+1));
}

function translatedChunkName(chunk)
{
	//return (chunk.position[0] - 303792) + "|0|" + (chunk.position[2] - 438258)
    return (chunk.position[0] - 294668) + "|0|" + (chunk.position[2] - 424954)
}

function calcChunkCoordsFromCoords(xlat,ylon,dims)
{
	return {x:Math.floor(xlat / (dims+1)),y:0,z:Math.floor(ylon / (dims+1))};
}

function coordArrayToChunks2(coordArray,roadname)
{

	//console.log(coordArray[0] , 'to' , coordArray[coordArray.length-1])
	var chunks = new Array();
	var chunkCoords = new Array();
	var voxels = new Int8Array(32 * 32 * 32);
	for (v=0;v < voxels.length;v++)
	{
		voxels[v] = 0;
	}	
	//console.log(coordArray.length, ' line points to parse.')
	for (i =0;i < coordArray.length;i++)
	{
		var x = coordArray[i].x 
		var y = coordArray[i].y 
		var chunkc = calcChunkCoordsFromCoords(x,y,32)
		var chunkname = calcChunkNameFromCoords(x,y,32)
		if (chunkname in chunks)
		{}
		else
		{
			chunks[chunkname] = 
			{
			roadname: roadname,
		    position : [chunkc.x,chunkc.y,chunkc.z],
			dims : [32,32,32],
			voxels : new Int8Array(32*32*32)
			}
			for(v1=0;v1<31;v1++){
				for(v2 = 0;v2 < 31;v2++){
					chunks[chunkname].voxels[(32*32*v1)+v2] = 1
				}
			}

		}
			var px,py,pz,pos
			px = x - (chunkc.x * (32 + 1))
			py = (chunkc.y * (32 + 1))
			pz = y - (chunkc.z * (32 + 1))
			pos = (32 * 32 * pz) + 0 + (px)
			chunks[chunkname].voxels[pos] = 2
			//fill sides of road center
			if (chunks[chunkname].voxels[pos-2] != 2) {chunks[chunkname].voxels[pos-2] = 3}
			if (chunks[chunkname].voxels[pos-1] != 2) {chunks[chunkname].voxels[pos-1] = 3}
			if (chunks[chunkname].voxels[pos+1] != 2) {chunks[chunkname].voxels[pos+1] = 3}
				if (chunks[chunkname].voxels[pos+2] != 2) {chunks[chunkname].voxels[pos+2] = 3}
			if (chunks[chunkname].voxels[(32*32)+pos] !=2) {chunks[chunkname].voxels[(32*32)+pos]= 3}
			if (chunks[chunkname].voxels[pos-(32*32)] !=2) {chunks[chunkname].voxels[pos-(32*32)] = 3}
			if (chunks[chunkname].voxels[(32*32*2)+pos] !=2) {chunks[chunkname].voxels[(32*32*2)+pos]= 3}
			if (chunks[chunkname].voxels[pos-(32*32*2)] !=2) {chunks[chunkname].voxels[pos-(32*32*2)] = 3}
	}
	return chunks;
}

function parseOSMtoVoxeljs(road){
//var chunks = coordArrayToChunks();
var chunks = new Array()// = coordArrayToChunks2();  //master array to hold chunks.
console.time('All-Meters');//Start a timer to track how long convertion from geojson to 
var polygons ={}
	for(i2 = 0;i2 < road.rowCount;i2++) //for every road in array/dataset
	{
		//i2 = road.features.length
		  //  	console.log(result.rows[i].row_to_json.geometry.coordinates);
    	//console.log(result.rows[i].row_to_json.properties)
		console.log(i2,"of",road.rowCount) //progress - how many roads do we have to left to process?
		var tempchunks = coordArrayToChunks2(parseLinestringRoadtoLineCoords(road.rows[i2].row_to_json.geometry.coordinates),road.rows[i2].row_to_json.properties.name)//processing functions - return assigned to temp varible.
		//console.log(road.features[i2].properties.name)
		for (var c in tempchunks)//check each chunk that was just spit out from processing functions....
		{
			//console.log(c.position)
			//process.exit();
			if (c in chunks)  //if a particular new chunk, already exists in the master array(in the same 3d space)..
			{
				//run logic to merge the new chunk to the existing chunk.  Don't overwrite air blocks?
				for(m = 0;m< (32*32*32);m++)
				{
					if(tempchunks[c].voxels[m] != 0)
					{
						chunks[tempchunks[c].position.join("|")].voxels[m] = tempchunks[c].voxels[m]
					}
				}

				chunks[tempchunks[c].position.join("|")].roadname += tempchunks[c].roadname
			}
			else
			{
			//if a particular chunk doesn't exist in the master array, add it to the master array.
			chunks[tempchunks[c].position.join("|")] = tempchunks[c]	
			}
			
		}

		//chunks.push.apply(chunks,coordArrayToChunks2(parseLinestringRoadtoLineCoords(road.features[i2].geometry.coordinates)))
	}

return chunks;
//console.log(Object.getOwnPropertyNames(chunks).sort());
}

function bytesToSize(input, precision)
{
    var unit = ['', 'K', 'M', 'G', 'T', 'P'];
    var index = Math.floor(Math.log(input) / Math.log(1024));
    if (unit >= unit.length) return input + ' B';
    return (input / Math.pow(1024, index)).toFixed(precision) + ' ' + unit[index] + 'B'
}

function displayMemUsage(){
var usage;
usage = process.memoryUsage();
console.log('RSS: ' + bytesToSize(usage.rss, 3), 'and Heap:', bytesToSize(usage.heapUsed, 3), 'of', bytesToSize(usage.heapTotal, 3), 'total');
console.timeEnd('All-Meters');
}

//vvvvvvvvvvvvvvvvvvvWrite calculations to MongoDBvvvvvvvvvvvvv

//process.exit();

// Example of a simple findOneAndReplace operation
function writeMongo(chunk){
//var chunks = new Array();
var chunkname = chunk.position.join("|");
/*
chunkname = '1|1|1'
chunks[chunkname] =
{
	//roadname: roadname,
    position : [1,1,1],
	dims : [32,32,32],
	voxels : new Int8Array(32*32*32)
}
*/
var MongoClient = require('mongodb').MongoClient,
  test = require('assert');
MongoClient.connect('mongodb://localhost:27017/osblocks', function(err, db) {
  // Get the collection
  var col = db.collection('chunks');
    col.findOneAndReplace({cname:chunkname}
      , {chunk:chunks[chunkname].voxels,cname:chunkname}
      , {
      	            projection: null
          , sort: null
          , returnOriginal: false
          , upsert: true
        }
      , function(err, r) {
        test.equal(null, err);
        test.equal(1, r.lastErrorObject.n);
        //test.equal(1, r.value.cname);

        db.close();
    });
});
}

File: OSVParser/parser.cpp
//Compile with:
//export PKG_CONFIG_PATH=/usr/local/lib/pkgconfig
//c++ --std=c++11 parser.cpp -o parser.out -L/usr/local/lib -lpq $(pkg-config --cflags --libs libmongocxx) && ./parser.out
//run with:
//./parser.out
/*
this link has an INVALID use of the driver.  Don't use it.
https://docs.mongodb.com/getting-started/cpp/update/
Use this instead:
https://github.com/mongodb/mongo-cxx-driver/blob/master/examples/mongocxx/update.cpp
thanks article: http://stackoverflow.com/questions/38845214/mongodb-c-tutorial-program-fails-mongocxxv-noabilogic-error
*/

#include <stdio.h>
#include <string>
#include <iostream>
#include <libpq-fe.h>
#include <math.h>
#include <vector>
#include <sstream>

#include <list>

#include <bsoncxx/builder/stream/document.hpp>
#include <bsoncxx/json.hpp>

//#include <mongocxx/options/find_one_and_replace.hpp>
#include <mongocxx/client.hpp>
#include <mongocxx/stdx.hpp>
#include <mongocxx/instance.hpp>
#include <mongocxx/uri.hpp>
using bsoncxx::builder::stream::document;
using bsoncxx::builder::stream::open_document;
using bsoncxx::builder::stream::close_document;
using bsoncxx::builder::stream::open_array;
using bsoncxx::builder::stream::close_array;
using bsoncxx::builder::stream::finalize;


struct point {
  int x;
  int y;
};

struct voxelxyz {
  int x;
  int y;
  int z;
  std::string nameFromXYZ()
  {
    return std::to_string(x) + '|' + std::to_string(y)  + '|' + std::to_string(z);
  }
};

struct chunk {
  std::string roadname;
  voxelxyz position;
  int dims[3];
  std::vector<int> voxels;
};


point convertEPSG3857toScreenXY(float lat,float lon)
{
  //todo: code for all 4 regions of global map.
  point p;
  p.x = 20026376 + floor(lat);
  p.y = 20048966 - floor(lon);
  return p;
}

std::vector<point> calcStraightLine (point startCoordinates, point endCoordinates)
{
    //puts("calcStraightLine");
    std::vector<point> coordinatesArray;
    // Translate coordinates
    int x1 = startCoordinates.x;
    int y1 = startCoordinates.y;
    int x2 = endCoordinates.x;
    int y2 = endCoordinates.y;
    // Define differences and error check
    int dx = abs(x2 - x1);
    int dy = abs(y2 - y1);
    int sx = (x1 < x2) ? 1 : -1;
    int sy = (y1 < y2) ? 1 : -1;
    int err = dx - dy;
    // Set first coordinates
    point p;
    p.x = x1;
    p.y = y1;
    coordinatesArray.push_back(p);
    //return coordinatesArray;
    // Main loop
    while (!((x1 == x2) && (y1 == y2))) {
      int e2 = err << 1;
      if (e2 > -dy) {
        err -= dy;
        x1 += sx;
      }
      if (e2 < dx) {
        err += dx;
        y1 += sy;
      }
      // Set coordinates
      point p;
      p.x = x1;
      p.y = y1;
      coordinatesArray.push_back(p);
      //console.log(x1,'+',y1);
    }
    // Return the result
    return coordinatesArray;
}





//Loop through linestring with points that should form a line.
//convert each point from gps to xy
//form line of xy's between each xy.
//combine all those lines together.
std::vector<point> parseLinestringRoadtoLineCoords(std::vector<point> LineStringArray)
{
  //puts("parseLinestringRoadtoLineCoords");
	std::vector<point> lineCoordsArray;
	for (int i = 0;i < LineStringArray.size() - 1;i++)
	{
    //ref#http://stackoverflow.com/questions/3177241/what-is-the-best-way-to-concatenate-two-vectors
		std::vector<point> B = calcStraightLine(
		 convertEPSG3857toScreenXY(LineStringArray[i].x,LineStringArray[i].y),
		 convertEPSG3857toScreenXY(LineStringArray[i+1].x,LineStringArray[i+1].y)
   );
    lineCoordsArray.reserve( lineCoordsArray.size() + B.size() ); // preallocate memory
    lineCoordsArray.insert( lineCoordsArray.end(), B.begin(), B.end() );
	}
	//console.log(lineCoordsArray);
	return lineCoordsArray;
}

voxelxyz calcChunkCoordsFromCoords(int xlat,int ylon,int dims)
{
  voxelxyz pos;
  pos.x = floor(xlat / (dims+1));
  pos.y = 0;
  pos.z = floor(ylon / (dims+1));
  return pos;
}

std::string calcChunkNameFromCoords(int xlat,int ylon,int dims)
{
	return std::to_string(floor(xlat / (dims+1))) + "|0|" + std::to_string(floor(ylon / (dims+1)));
}


std::vector<chunk> coordArrayToChunks2(std::vector<point> coordArray, std::string roadname)
{
  //puts("coordArrayToChunks2");
	//console.log(coordArray[0] , 'to' , coordArray[coordArray.length-1])
	std::vector<chunk> chunks;
	std::vector<int> voxels(32 * 32 * 32);
	for (int v=0;v < voxels.size();v++)
	{
		voxels[v] = 0;
	}
  for(int v1=0;v1<31;v1++){
    for(int v2 = 0;v2 < 31;v2++){
      voxels[(32*32*v1)+v2] = 1;
    }
  }
	//console.log(coordArray.length, ' line points to parse.')
	for (int i =0;i < coordArray.size();i++)
	{
		int x = coordArray[i].x;
		int y = coordArray[i].y;
		voxelxyz chunkc = calcChunkCoordsFromCoords(x,y,32);
    int currentchunkindex = 0;
		//var chunkname = calcChunkNameFromCoords(x,y,32)
    bool found = false;
    for(int c=0;c<chunks.size();c++){
      if(chunks[c].position.x == chunkc.x && chunks[c].position.y == chunkc.y && chunks[c].position.z == chunkc.z){
        found = true;
        currentchunkindex = c;
        break;
      }
    }
		if (!found)
		{

      //console.log('no exist', chunkname)
      chunk tempchunk;
      tempchunk.roadname = roadname;
      tempchunk.position = chunkc;
      tempchunk.dims[0] = 32;
      tempchunk.dims[1] = 32;
      tempchunk.dims[2] = 32;
      tempchunk.voxels = voxels;
      chunks.push_back (tempchunk);
      currentchunkindex = chunks.size() - 1;
      //printf("%d",chunks.size());
		}
    //puts("try");
    int px = x - (chunkc.x * (32 + 1));
    int py = (chunkc.y * (32 + 1));
    int pz = y - (chunkc.z * (32 + 1));
    int pos = (32 * 32 * pz) + 0 + px;
    //std::cout << found << std::endl;
    //printf("%d,%d\n",currentchunkindex,pos);
    //printf("index = %d\n",currentchunkindex);
    //printf("%d %d %d %d %d %d %d %d %d %d",pos,pos-2,pos-1,pos+1,pos+1,pos+2,(32*32)+pos,pos-(32*32),(32*32*2)+pos,pos-(32*32*2));
    if(pos >= 0 && pos < (32*32*32)){
      //printf("%d\n",chunks.size());
      //printf("%d\n",chunks[currentchunkindex].voxels.size());
    chunks[currentchunkindex].voxels[pos] = 2;}
    //puts("success");
    //fill sides of road center
    if (pos-2 >= 0 && pos-2 < (32*32*32)) {if (chunks[currentchunkindex].voxels[pos-2] != 2) {chunks[currentchunkindex].voxels[pos-2] = 3;}}
    if (pos-1 >= 0 && pos-1 < (32*32*32)) {if (chunks[currentchunkindex].voxels[pos-1] != 2) {chunks[currentchunkindex].voxels[pos-1] = 3;}}
    if (pos+1 >= 0 && pos+1 < (32*32*32)) {if (chunks[currentchunkindex].voxels[pos+1] != 2) {chunks[currentchunkindex].voxels[pos+1] = 3;}}
    if (pos+2 >= 0 && pos+2 < (32*32*32)) {if (chunks[currentchunkindex].voxels[pos+2] != 2) {chunks[currentchunkindex].voxels[pos+2] = 3;}}
    if ((32*32)+pos >= 0 && (32*32)+pos < (32*32*32)) {if (chunks[currentchunkindex].voxels[(32*32)+pos] !=2) {chunks[currentchunkindex].voxels[(32*32)+pos]= 3;}}
    if (pos-(32*32) >= 0 && pos-(32*32) < (32*32*32)) {if (chunks[currentchunkindex].voxels[pos-(32*32)] !=2) {chunks[currentchunkindex].voxels[pos-(32*32)] = 3;}}
    if ((32*32*2)+pos >= 0 && (32*32*2)+pos < (32*32*32)) {if (chunks[currentchunkindex].voxels[(32*32*2)+pos] !=2) {chunks[currentchunkindex].voxels[(32*32*2)+pos]= 3;}}
    if (pos-(32*32*2) >= 0 && pos-(32*32*2) < (32*32*32)) {if (chunks[currentchunkindex].voxels[pos-(32*32*2)] !=2) {chunks[currentchunkindex].voxels[pos-(32*32*2)] = 3;}}
	}
  //console.log('chunk', chunks['294591|0|424978'])
  //puts("hmms");
	return chunks;
}

std::vector<point> parselinestringtovector(std::string linestring){
  //puts("parselinestringtovector");
  //ref#http://stackoverflow.com/questions/1195675/convert-a-char-to-stdstring
  //ref#http://stackoverflow.com/questions/1894886/parsing-a-comma-delimited-stdstring
  std::string localls = linestring;
  std::vector<point> pointarr;
  localls.erase(0,11);
  localls.erase(localls.length()-1);
  //std::cout << localls << std::endl;
  std::stringstream ss(localls);
  std::vector<std::string> phase1;
  while( ss.good() )
  {
      std::string substr;
      getline( ss, substr, ',' );
      phase1.push_back( substr );
      //puts("hmm");
  }
  for(int i = 0;i<phase1.size();i++){
    //printf("%s",phase1[i]);
    //std::cout << phase1[i] << std::endl;
    point p;
    std::stringstream ss2(phase1[i]);
    std::vector<std::string> temp;
    while( ss2.good() )
    {
        std::string substr;
        getline( ss2, substr, ' ' );
        temp.push_back( substr );
    }
    //std::cout << temp[0] << std::endl;
    //std::cout << temp[1] << std::endl;
    p.x = atoi(temp[0].c_str());
    p.y = atoi(temp[1].c_str());
    //printf("%d %d\n", p.x,p.y);
    pointarr.push_back(p);
  }

  return pointarr;
}


int main()
{
 std::cout << "Hello world!" << std::endl;
 PGconn          *pgconn;
 PGresult        *res;
 int             rec_count;
 int             row;
 int             col;
 mongocxx::instance inst{};
 mongocxx::client conn{mongocxx::uri{}};
 auto db = conn["osblocks"];
 //mongocxx::collection collection = conn["osblocks"]["chunks"];
 pgconn = PQconnectdb("dbname=gis host=localhost user=postgres password=Asdfjkl1");
 if (PQstatus(pgconn) == CONNECTION_BAD) {
  puts("We were unlineCoordsArrayle to connect to the datlineCoordsArrayase");
  exit(0);
 }

 //res = PQexec(conn, "declare geo no scroll cursor with hold for SELECT row_to_json(fc) from (select ST_AsGeoJSON(lg.way)::json As geometry, row_to_json(lp) As properties FROM planet_osm_line As lg INNER JOIN (SELECT osm_id, name,highway FROM planet_osm_line where highway is not null) As lp ON lg.osm_id = lp.osm_id) as fc;");
 res = PQexec(pgconn, "SELECT ST_AsText(way),osm_id, name,highway FROM planet_osm_line where highway is not null;");
 //res = PQexec(conn, "select * from planet_osm_line;");

 if (PQresultStatus(res) != PGRES_TUPLES_OK) {
  puts("We did not get any data!");
  exit(0);
 }
 rec_count = PQntuples(res);
 printf("We received %d records.\n", rec_count);
 puts("==========================");
 for (row=0; row<rec_count; row++) {
   printf("Record# %d\n",row);
  //for (col=0; col<4; col++) {
   //printf("%s\t\n", PQgetvalue(res, row, col));
   //parselinestringtovector(std::string(PQgetvalue(res, row, 0)))
   //break;
   std::vector<chunk> c;

   /*
   printf("%d",
   coordArrayToChunks2(
    parseLinestringRoadtoLineCoords(
       parselinestringtovector(std::string(PQgetvalue(res, row, 0)))),
    std::string(PQgetvalue(res, row, 2))).size()
  );
  */
  c = coordArrayToChunks2(
   parseLinestringRoadtoLineCoords(
      parselinestringtovector(std::string(PQgetvalue(res, row, 0)))),
   std::string(PQgetvalue(res, row, 2)));
   for(int i=0;i<c.size();i++)
  {
    std::cout << c[i].position.nameFromXYZ() << std::endl;

//    std::list<int> aList(c[i].voxels.begin(),c[i].voxels.end());
//BSONArrayBuilder arrBuilder;
//arrBuilder.append(aList);
//bsoncxx::BsonArray barr;

    bsoncxx::builder::stream::document filter_builder;//,
    document replace_builder;

filter_builder << "cname"
               << "294591|0|424978";

replace_builder << "cname"
                << c[i].position.nameFromXYZ();
                //<< "294591|0|424978";
    auto replace_builder2 = replace_builder
                << "chuck"
                << open_array;
                //<< c[i].voxels
                for(int n=0;n<c[i].voxels.size();n++){
                  replace_builder2 << c[i].voxels[n];
                }

                replace_builder2 << close_array;
                bsoncxx::document::value doc = replace_builder << finalize;
                //std::cout << bsoncxx::to_json(doc) << std::endl;
                //<< finalize;

//db["chunks"].options.upsert(true);
mongocxx::options::update u{};


u.upsert(true);
db["chunks"].replace_one(filter_builder.view(), doc.view(),u);//,u.upsert(true));
//collection.replace_one(filter_builder.view(), replace_builder.view(),u);//,u.upsert(true));

/**/
  }
   //writeMongo(coordArrayToChunks2)
  //}
  puts("");
 }
 puts("==========================");
 int wait;
 std::cin >> wait;
 PQclear(res);
 PQfinish(pgconn);
 //point p = convertEPSG3857toScreenXY(-10301828.82, 6037668.61);
 //point p2 = convertEPSG3857toScreenXY(-10301855.73, 6037449.53);
 //std::vector<point> v = calcStraightLine(p,p2);
 //for(int i=0;i<v.size();i++){
//   printf("coord %d %d \n",v[i].x,v[i].y);
 //}
 //printf("%d",p.x);
 return 0;
}
